### PR TITLE
aarch64: rationalize exception entry/exit code

### PR DESCRIPTION
**Pull request by npitre**

### CODE REVIEW

File: arch/arm/core/aarch64/vector_table.S [hunk 1]
```diff
@@ -62,11 +102,12 @@ SECTION_SUBSEC_FUNC(exc_vector_table,_vector_table_section,_vector_table)
 
 	/* Current EL with SP0 / Synchronous */
 	.align 7
+	z_arm64_enter_exc x0, x1
```
Comment (carlocaione) [inline arch/arm/core/aarch64/vector_table.S:105]: I'm a bit torn about this change. I understand the rationale behind this change but I cannot say I particularly like it.

- First of all I understand that the most logical place to have `z_arm64_exit_exc` / `z_arm64_enter_exc` is in the vector table but they do look out of place (at least IMO)
- With this change you need to have one as a macro, the other as a function that is quite confusing if you don't know the details
- You are using `z_arm64_enter_exc` as preamble for each vector but there is no corresponding `z_arm64_exit_exc` that is still in the out-of-line code

So at the end of the day I wonder if this change is really better than having to waste space in each vector slot. All the remaining fixes are really good though.

**Comment by npitre** [comment]
On Wed, 24 Feb 2021, Carlo Caione wrote:

> - First of all I understand that the most logical place to have `z_arm64_exit_exc` / `z_arm64_enter_exc` is in the vector table but they do look out of place (at least IMO)

It is logical... or it is out of place. You can't have both in the same sentence.  ;-)
We can discuss the naming if that might make you more comfortable, but IMHO those pieces clearly belong to that file.

> - With this change you need to have one as a macro, the other as a function that is quite confusing if you don't know the details

If one gets that confused here then maybe they should consider a career change.
There are far more trickier pieces in Zephyr in order to save much fewer bytes already. In comparison, this change is a marvell of clarity.

Seriously, again I kept the symmetry in the naming to make it clearer. It's a bad sign if that didn't work on you so let's find better names..

> - You are using `z_arm64_enter_exc` as preamble for each vector but there is no corresponding `z_arm64_exit_exc` that is still in the out-of-line code

On purpose, for a few reasons: vector slots are a bit tight to also contain the exception return code, the exception routines would have to also preserve x30 across multiple bl calls which would grow the code and stack usage, and if ever the last vector slot is used one day then it will push down the following code with the same offset as the size of the single return code instance with this patch anyway. In the end what I propose here is the best compromise so far.

> So at the end of the day I wonder if this change is really better than having to waste space in each vector slot.

Oh absolutely. We can work on the estetics, but there is no denying possible on the technical side.

**Comment by carlocaione** [comment]
> It is logical... or it is out of place. You can't have both in the same sentence. ;-)

ah, fair enough :) Let me rephrase then: my point is that `vector_table.S` was supposed to only contain the vector table indeed but now it is containing a macro (while usually the macros are in a separate `inc` file) and a function that is not even used in this file.

> We can discuss the naming if that might make you more comfortable, but IMHO those pieces clearly belong to that file.

nah, naming is fine I guess because the symmetry is clear enough. 

> On purpose, for a few reasons: vector slots are a bit tight to also contain the exception return code, the exception routines would have to also preserve x30 across multiple bl calls which would grow the code and stack usage, and if ever the last vector slot is used one day then it will push down the following code with the same offset as the size of the single return code instance with this patch anyway. In the end what I propose here is the best compromise so far.

Yeah, no questioning the technical opportunity to do that. It's again the imbalance that is disturbing me.

> Oh absolutely. We can work on the estetics, but there is no denying possible on the technical side.

No doubt but not sure what we can do about the esthetics in this case.

File: arch/arm/core/aarch64/vector_table.S [hunk 2]
```diff
@@ -13,10 +13,50 @@
...
+.macro z_arm64_enter_exc xreg0, xreg1
+
+	/* Switch to SP_EL0 */
+	msr	spsel, #0
+
+	/*
+	 * Two things can happen to the remaining registers:
+	 *
+	 * - No context-switch: in this case x19-x28 are callee-saved register
+	 *   so we can be sure they are not going to be clobbered by ISR.
```
Comment (jharris-intel) [inline arch/arm/core/aarch64/vector_table.S:36]: As an aside: in my experience this is a very good invariant to add a debug check for (later, not in this PR!).

Even just storing the XOR of the callee-save registers and checking it at the end is rather helpful.

*****

There are way too many things that can result in callee-save registers being clobbered, and the resulting debugs are... well, I find them fun, but I'm pretty sure I'm the outlier there.

File: arch/arm/core/aarch64/vector_table.S (hunk 1 above)
Comment (jharris-intel) [reply arch/arm/core/aarch64/vector_table.S:105]: I think I'm forced to agree with you for the most part here, mainly because this isn't _actually_ saving space, because the vectors are already padded out.

Perhaps have both the entry and exit code in each of the vector slots?

The only space you could save here compared to having entry and exit in the vector slots is after the last vector slot in the table - but then you're replacing one exception exit with one exception exit, which isn't saving space. (Actually likely making the total space use _worse_, because the last vector is just a `b .` with no exception exit...)

It does mean that the actual exception handlers (`z_arm64_serror`, `_isr_wrapper`, etc) need to save/restore their LR, which isn't perfect.

**Comment by npitre** [comment]
On Wed, 24 Feb 2021, jharris-intel wrote:

> I think I'm forced to agree with you for the most part here, mainly because this isn't _actually_ saving space, because the vectors are already padded out.

Not sure I understand you here.

> Perhaps have both the entry and exit code in each of the vector slots?

As I said previously, there's room for 32 instructions only. The context 
saving already takes up 15 instructions. The context restoring would 
take another 15 instructions. That's if we don't change anything (SMP 
might bring some more stuff in here -- I didn't look too closely). That 
doesn't leave much slack once you put the required bl in.

Then you have to save x30 before calling any other functions. That's 16 
bytes of stack usage that are unnecessary otherwise, as well as the 
additional str/ldr themselves.

> The only space you could save here compared to having entry and exit in the vector slots is after the last vector slot in the table - but then you're replacing one exception exit with one exception exit, which isn't saving space. (Actually likely making the total space use _worse_, because the last vector is just a `b .` with no exception exit...)

Right. And even if there was something to do in the last slot, it would 
need its exception exit too. So having it out-of-line can take the same 
space as if it were in the last slot.

**Comment by jharris-intel** [comment]
@npitre -

> That's 16 bytes of stack usage that are unnecessary otherwise, as well as the additional str/ldr themselves.

I mean, you don't need to save it on the stack (and indeed I'd recommend not saving it on the stack). You can move it to a callee-save register I think (you just saved context anyway). But yes, it does add unnecessary instructions.

> As I said previously, there's room for 32 instructions only. The context saving already takes up 15 instructions. The context restoring would take another 15 instructions.

Right. I'm way too used to thumb-2 for my guesstimates of code size (both LDM/STM and that many instructions are 16 bits - overall thumb-2 exception entry is a lot smaller.).

Given that you'd likely have to split the exception elsewhere, I do see the appeal of tail-calling to a shared exception return.

> Not sure I understand you here.

If you could fit the exception entry and exit into the vector table entries (...which is largely a moot point, because as you mention you likely cannot), having exception entry and exit in each entry looks roughly like:

```
    offset 0:
        exception entry
        branch
        exception exit
        <little bit of padding>
    offset 0x80
        exception entry
        branch
        exception exit
        <little bit of padding>
    offset 0x100
        exception entry
        branch
        exception exit
        <little bit of padding>
     ...
    offset 0x780
        exception entry
        branch
        exception exit
```

This in total takes up `0x780 + sizeof(exception entry) + sizeof(branch) + sizeof(exception exit)` bytes of space.

```
    offset 0:
        exception entry
        branch
        <more padding>
    offset 0x80
        exception entry
        branch
        <more padding>
    offset 0x100
        exception entry
        branch
        <more padding>
     ...
    offset 0x780
        exception entry
        branch
    shared_exit:
        exception exit
```

This in total takes up ... `0x780 + sizeof(exception entry) + sizeof(branch) + sizeof(exception exit)` bytes of space. (Give or take.)

(As an aside, this is one of the reasons I prefer AArch32's method here. AArch64's exception entry ends up taking up a lot of space for what ends up in practice being pretty much the same 'branch somewhere else immediately' as AArch32 does...)

**Comment by npitre** [comment]
On Wed, 24 Feb 2021, jharris-intel wrote:

> I mean, you don't need to save it on the stack (and indeed I'd recommend not saving it on the stack). You can move it to a callee-save register I think (you just saved context anyway).

Nope. Yes we saved context but only for the otherwise clobberable set. 
So the actual functions we call will take care of preserving the 
callee-saved registers if they need them. Additionally we do save the 
callee-saved registers only when performing an actual task switch. But 
on anything but a task switch that would be some unnecessary overhead.

In other words, before moving x30 to a callee-saved register, we would 
have to preserve that callee-saved register in the first place. Better 
store x30 on the stack ourselves.

But given that all exceptions have the same exit code, we actually don't 
need to remember x30 before calling any other function if that code is 
fatored out. We can just branch to it directly.

> (As an aside, this is one of the reasons I prefer AArch32's method here. AArch64's exception entry ends up taking up a lot of space for what ends up in practice being pretty much the same 'branch somewhere else immediately' as AArch32 does...)

I agree.

To be fair it is nice to be able to save your context and then branch to 
separate areas depending on the exception. But in the end, it is 
probably better to just duplicate your context saving code in every 
exception handler like on Aarch32 which footprint wise is still a win.

And the FIQ vector, the only one with justification for possibly very 
tight latency constraints, is the very last so you can put your FIQ code 
right there if the price of an extra branch is too much.

**Review by sandeepbrcm — APPROVED** [review]
functionality wise this is the best that can be achieved w.r.t space optimization, keeping the entry save instructions inline and leaving some head room for future expansion (just in case someday something more needs to be done in entry path). If you had somehow packed the exit routine as well then that would have some concern.

I agree this might reduce the readability slightly, but someone looking at `vactor_table.S` can't avoid details anyway!

**Review by carlocaione — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
